### PR TITLE
Refactor README.md

### PR DESCRIPTION
**Pull request by mkgrgis**
By https://github.com/ibarwick/firebird_fdw/blob/master/README.md example of advanced FDW documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,16 +1,87 @@
-# GridDB Foreign Data Wrapper for PostgreSQL
+GridDB Foreign Data Wrapper for PostgreSQL
+==========================================
 
-This PostgreSQL extension is a Foreign Data Wrapper (FDW) for [GridDB][1].  
-This version of griddb_fdw can work for PostgreSQL 11, 12, 13, 14 and 15. It is confirmed in GridDB 5.0.0.
```
Comment (hrkuma) [inline README.md:4]: I think the information of supported version is important, so could you keep this information?

File: README.md [hunk 2]
```diff
@@ -19,192 +90,345 @@ $ make
...
+- **use_remote_estimate** as *boolean*, optional
+
+- **batch_size** as *integer*, optional
+
+- **fdw_startup_cost** as *interger*, optional
+
+- **fdw_tuple_cost** as *integer*, optional
 
-### update and delete
+- **rowkey** as *string*, optional
```
Comment (hrkuma) [inline README.md:166]: Could you remove this rowkey in this section as TABLE options?

File: README.md [hunk 3]
```diff
@@ -19,192 +90,345 @@ $ make
...
-    * For the push down of now() and timestamp() function, please add prefix `griddb_` for these functions.     
-      Example: now() -> griddb_now()
+### Grant usage on foreign server to normal user in PostgreSQL:
 
-- Support LIMIT...OFFSET pushdown when having LIMIT clause only or both LIMIT and OFFSET.
-- Support ORDER BY pushdown with column, except ORDER BY column ASC NULLS FIRST, ORDER BY column DESC NULLS LAST and ORDER BY functions/formulas
-- Support aggregation push down:
-    * Common functions of Postgres and Griddb are pushed down: min, max, count(*), sum, avg, variance, stddev.
-    * Unique function of Griddb are pushed down: time_avg<br>
+Once for a normal user (non-superuser) in PostgreSQL, as PostgreSQL superuser. It is a good idea to use a superuser only where really necessary, so let's allow a normal user to use the foreign server (this is not required for the example to work, but it's secirity recomedation).
```
Comment (hrkuma) [inline README.md:246]: Please fix the typo 'secirity'?

File: README.md [hunk 4]
```diff
@@ -19,192 +90,345 @@ $ make
...
+	SERVER griddb_svr 
+    	OPTIONS (
+	  username 'username',
+	  password 'password'
+	);
+```
+Where `pguser` is a sample user for works with foreign server (and foreign tables).
+
+### Create foreign table
+All `CREATE FOREIGN TABLE` SQL commands can be executed as a normal PostgreSQL user if there were correct `GRANT USAGE ON FOREIGN SERVER`. No need PostgreSQL supersuer for secirity reasons but also works with PostgreSQL supersuer.
```
Comment (hrkuma) [inline README.md:268]: Please fix the typo 'secirity'?

File: META.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+         "abstract": "Foreign Data Wrapper for GridDB databases",
+         "file": "griddb_fdw.c",
+         "docfile": "README.md",
+         "version": "0.0.1"
+      }
+   },
+   "prereqs": {
+      "runtime": {
+         "requires": {
+            "PostgreSQL": "10.0.0"
```
Comment (hrkuma) [inline META.json:19]: This FDW supports from version 11.
So clould you update the PostgreSQL required version?

File: META.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
+{
+   "name": "griddb_fdw",
+   "abstract": "Foreign Data Wrapper for GridDB databases",
+   "description": "PostgreSQL extension which implements a Foreign Data Wrapper (FDW) for GridDB.",
+   "version": "0.0.1",
```
Comment (hrkuma) [inline META.json:5]: The latest version is 2.3.0.
So could you update this version?

File: META.json [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+   "maintainer": "pgspider",
+   "license": "",
+   "provides": {
+      "sgriddb_fdw": {
+         "abstract": "Foreign Data Wrapper for GridDB databases",
+         "file": "griddb_fdw.c",
+         "docfile": "README.md",
+         "version": "0.0.1"
```
Comment (hrkuma) [inline META.json:13]: The latest version is 2.3.0.
So could you update this version?

File: META.json [hunk 8]
```diff
@@ -0,0 +1,44 @@
+{
+   "name": "griddb_fdw",
+   "abstract": "Foreign Data Wrapper for GridDB databases",
+   "description": "PostgreSQL extension which implements a Foreign Data Wrapper (FDW) for GridDB.",
+   "version": "0.0.1",
+   "maintainer": "pgspider",
+   "license": "",
+   "provides": {
+      "sgriddb_fdw": {
```
Comment (hrkuma) [inline META.json:9]: Could you fix this as "griddb_fdw"?

**Review by hrkuma — CHANGES_REQUESTED** [review]
Thank you for fixing proposal.
I have commented some.
Could you check them?

File: README.md (hunk 2 above)
Comment (mkgrgis) [reply README.md:166]: Ok. I think descriptions by authors for all options will be usefully. During studying `gridsb_fdw` I don't understand some options.

File: README.md (hunk 1 above)
Comment (mkgrgis) [reply README.md:4]: Added again, @hrkuma.

File: README.md (hunk 3 above)
Comment (mkgrgis) [reply README.md:246]: Fixed, thanks @hrkuma.

File: README.md (hunk 4 above)
Comment (mkgrgis) [reply README.md:268]: Done. "u" and "i" is near on keyboard ...

File: META.json (hunk 8 above)
Comment (mkgrgis) [reply META.json]: Fixed.

File: META.json (hunk 7 above)
Comment (mkgrgis) [reply META.json]: Added.

File: META.json (hunk 6 above)
Comment (mkgrgis) [reply META.json:5]: Updated.

File: META.json (hunk 5 above)
Comment (mkgrgis) [reply META.json]: Fixed.

**Comment by mkgrgis** [comment]
> Could you check them?

Yes, thanks @hrkuma! I think all is fixed.

File: README.md [hunk 9]
```diff
@@ -1,210 +1,440 @@
...
+Identifier case handling
+------------------------
+
+PostgreSQL folds identifiers to lower case by default.
+Rules and problems with griddb identifiers **yet not tested and described**.
+
+Generated columns
+-----------------
+
+Behavoiur within generated columns **yet not tested and described**.
```
Comment (nxhai98) [inline README.md:207]: `generated columns` has been supported and tested, here is test result: https://github.com/pgspider/griddb_fdw/blob/ae1c1632558d702ce21d659a071a32d06c240b5c/expected/15.0/griddb_fdw_post.out#L7898

So, could you add this information?

File: README.md [hunk 10]
```diff
@@ -1,210 +1,440 @@
...
+```sql
+INSERT INTO ft1(c1, c2) VALUES(11, 12) ON CONFLICT DO NOTHING;
+```
+
+- Don't support `RETURNING`. Returning is way to obtain data if rows are modified by `INSERT`, `UPDATE`, and `DELETE` commands. `griddb_fdw` does not support this feature.
+```sql
+INSERT INTO ft1 (c0, c1) VALUES (1, 2) RETURNING c0, c1;
+```
+- Don't support DIRECT MODIFICATION
+- Don't support `IMPORT FOREIGN SCHEMA` with option `import_generated` or `LIMIT TO` clause
```
Comment (nxhai98) [inline README.md:321]: This description seems be not correct:
- Griddb FDW makes CREATE FOREIGN TABLE statements for all existed tables in remote server, and then PostgreSQL's core checks LIMIT TO or EXCEPT conditions.
- `import_generated` is an option of postges_fdw and it has no meaning with griddb_fdw.

So, could you update this information?

I think the description should be:
```suggestion
- Don't support push down LIMIT TO | EXCEPT filter in IMPORT FOREIGN SCHEMA. These filters will run locally.
```

**Review by nxhai98 — COMMENTED** [review]
I have some comment for your fix.
Could you check them?

File: README.md (hunk 1 above)
Comment (hrkuma) [reply README.md:4]: Thank you for support.

File: README.md (hunk 2 above)
Comment (hrkuma) [reply README.md]: Thank you for support.

File: README.md (hunk 4 above)
Comment (hrkuma) [reply README.md]: Thank you for support.

File: META.json (hunk 6 above)
Comment (hrkuma) [reply META.json]: Thank you for support.

File: META.json (hunk 8 above)
Comment (hrkuma) [reply META.json]: Thank you for support.

File: META.json (hunk 7 above)
Comment (hrkuma) [reply META.json]: Thank you for support.

File: META.json (hunk 5 above)
Comment (hrkuma) [reply META.json]: Thank you for support.

File: README.md (hunk 3 above)
Comment (hrkuma) [reply README.md]: Thank you for support.

**Review by hrkuma — APPROVED** [review]

File: README.md (hunk 9 above)
Comment (mkgrgis) [reply README.md]: Yes, new information have been added. Please, check.

File: README.md (hunk 10 above)
Comment (mkgrgis) [reply README.md]: Thanks @nxhai98  , changed.

**Comment by mkgrgis** [comment]
> I have some comment for your fix. Could you check them?

Yes. Thanks, @nxhai98, all changes has been made.

File: README.md (hunk 9 above)
Comment (nxhai98) [reply README.md]: Thank you for supporting.

File: README.md (hunk 10 above)
Comment (nxhai98) [reply README.md]: Thank you for fixing,

**Review by hrkuma — CHANGES_REQUESTED** [review]

File: README.md [hunk 11]
```diff
@@ -111,11 +111,11 @@ Usage
 
   GridDB notification port.
 
-- **database** as *string*, optional
+- **database** as *string*, optional, default `public`
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply README.md:114]: I have found https://github.com/pgspider/griddb_fdw/blob/ae1c1632558d702ce21d659a071a32d06c240b5c/connection.c#L252

File: README.md [hunk 12]
```diff
@@ -111,11 +111,11 @@ Usage
...
 
-  GridDB database name. Default is `public` if skipped.
+  GridDB database name.
 
-- **clustername** as *boolean*, optional
+- **clustername** as *boolean*, optional, default `false`
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply README.md]: Thanks @hrkuma , this error has been changed to right sentence.

File: README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (hrkuma) [reply README.md:115]: Oh, I was wrong.
(I checked only option processing.)
Thank you for anyway.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (hrkuma) [reply README.md]: Thank you for support.

**Review by hrkuma — APPROVED** [review]

**Comment by hrkuma** [comment]
@mkgrgis
Thank you for improving documentation.  
I'd like to merge this.

**Comment by mkgrgis** [comment]
Thank you for carefully reading, @hrkuma !

**Pull request merged**